fish-rust/src/output.rs

```diff
@@ -8,12 +24,718 @@ bitflags! {
+    COLOR_SUPPORT.store(val, Ordering::Relaxed);
+}
+
+/// Returns true if we think tparm can handle outputting a color index.
+fn term_supports_color_natively(c: i32, term: &Term) -> bool {
+    #[allow(clippy::int_plus_one)]
+    if let Some(max_colors) = term.max_colors {
+        max_colors >= c + 1
+    } else {
+        false
+    }
+}
+
+pub fn get_color_support() -> ColorSupport {
+    let val = COLOR_SUPPORT.load(Ordering::Relaxed);
+    ColorSupport::from_bits_truncate(val)
+}
+
+pub fn set_color_support(val: ColorSupport) {
+    COLOR_SUPPORT.store(val.bits(), Ordering::Relaxed);
+}
+
+// These are historic. writembs() attempts to write a multibyte string of type &Option<CString> to the given outputter.
+// writembs() will noisily fail, writembs_nofail will not.
+macro_rules! writembs(
+    ($outp: expr, $mbs: expr) => {
+        crate::output::writembs_check($outp, $mbs, std::stringify!($mbs), true, std::file!(), std::line!())
+    }
+);
+
+macro_rules! writembs_nofail(
+    ($outp: expr, $mbs: expr) => {
+        crate::output::writembs_check($outp, $mbs, std::stringify!($mbs), false, std::file!(), std::line!())
+    }
+);
+pub(crate) use writembs_nofail;
+
+fn index_for_color(c: RgbColor) -> u8 {
+    if c.is_named() || !(get_color_support().contains(ColorSupport::TERM_256COLOR)) {
+        return c.to_name_index();
+    }
+    c.to_term256_index()
+}
+
+// Given a Cursor which have used write! on, return the slice of written bytes.
+fn cursor_to_slice(cursor: Cursor<&mut [u8]>) -> &[u8] {
+    let len: usize = cursor
+        .position()
+        .try_into()
+        .expect("cursor position overflow");
+    let buff: &mut [u8] = cursor.into_inner();
+    &buff[..len]
+}
+
+fn write_color_escape(
+    outp: &mut Outputter,
+    term: &Term,
+    todo: &CStr,
+    mut idx: u8,
+    is_fg: bool,
+) -> bool {
+    if term_supports_color_natively(idx.into(), term) {
+        // Use tparm to emit color escape.
+        writembs!(outp, tparm1(todo, idx.into()));
+        true
+    } else {
+        // We are attempting to bypass the term here. Generate the ANSI escape sequence ourself.
+        let mut buff = [0; 32];
+        let mut cursor = Cursor::new(&mut buff[..]);
+        if idx < 16 {
+            // this allows the non-bright color to happen instead of no color working at all when a
+            // bright is attempted when only colors 0-7 are supported.
+            //
+            // TODO: enter bold mode in builtin_set_color in the same circumstance- doing that combined
+            // with what we do here, will make the brights actually work for virtual consoles/ancient
+            // emulators.
+            if term.max_colors == Some(8) && idx > 8 {
+                idx -= 8;
+            }
+            write!(
+                cursor,
+                "\x1B[{}m",
+                (if idx > 7 { 82 } else { 30 }) + idx as i32 + ((!is_fg as i32) * 10)
+            )
+            .unwrap();
```

Comment:
> How does `write!` fail? Allocation failure?

[It can't](https://doc.rust-lang.org/std/primitive.slice.html#impl-Write-for-%26mut+%5Bu8%5D), but since `std::io::Write::write()` always returns a `Result<usize, std::io::Error>`, there's no way to signal this in the type system.

> So it's probably a good idea to use `.expect()` in most places so we can add a good error.

True, however in this specific case there is no good error message because it can never fail.

Reply:
I would use `.unwrap_or_else(|_| unreachable!())` or even `.unwrap_or_else(|_| unsafe { core::hint::unreachable_unchecked() })`.

Reply:
Personally I'm firmly against using any unnecessary unsafe code without prior benchmarks that show a clear performance improvement. There's no point, and it's way too easy to introduce UB (either immediately or through refactorings).

Reply:
I switched to `expect()` with a better error message.